iree/turbine/kernel/wave/wave.py

```diff
@@ -294,21 +306,42 @@ def is_reduction(node: fx.Node):
                 if tiling_constraint.dim == custom.axis:
                     tiling_constraint.induction_var = self.induction_vars[custom]
 
-    def initialize_wave_constraints(self, trace: CapturedTrace) -> None:
+    def initialize_implicit_constraints(self, trace: CapturedTrace) -> None:
         """
-        For each wave constraint, determines the appropriate wave id by looking
-        for workgroup constraints along the same dimension and using information
-        from the hardware constraints.
+        Initialize all implicit constraints. Add wave constraints to the
+        [x, y, z] dimensions using information from the hardware constraints.
 
         """
 
         hardware_constraint = self.hardware_constraints[0]
-        for wave_constraint in self.wave_constraints:
-            for workgroup_constraint in self.workgroup_constraints:
-                if wave_constraint.dim == workgroup_constraint.dim:
-                    wave_constraint.set_wave_id_from_hardware_and_workgroup_constraint(
-                        hardware_constraint, workgroup_constraint
-                    )
+        constraints = []
+        for workgroup_constraint in self.workgroup_constraints:
+            # Ignore `workgroup_dim > 2` and dimensions with only one 
+            # wave (ie. the value is always 0).
+            if (
+                workgroup_constraint.workgroup_dim > 2
+                or (
+                    workgroup_constraint.workgroup_dim != 0
+                    and hardware_constraint.waves_per_block[
+                        workgroup_constraint.workgroup_dim
+                    ]
+                    == 1
+                )
+                or not workgroup_constraint.primary
+            ):
+                continue
+            constraint = ImplicitWaveConstraint(
```

Comment:
I am a big fan of "explicitly" calling out that this is an implicit constraint and removing the footgun of 2 specifications that should agree but may sometimes not. I didn't dare going this path myself as I expected things to generally break. 

If we can get confirmation that this is indeed the intended semantics this would be fantastic and a really great catch!

Maybe let's phrase it differently, how would we expect the system to behave when the 2 don't agree and what are theoretical use cases? 

I personally have a lot of cases where mapping and number of threads/waves per block disagree but this involves non-trivial predications (up to an including wave-specialization); it seems like a big stretch to support these in TKW but maybe this is the intended future @harsh-amd ?